[flutter_tools] UnsupportedError at _Uri._checkWindowsPathReservedCharacters when parsing directory entries with invalid characters in `assets` section in pubspec.yaml

On `beta/3.18.0-0.1.pre`.
Command: `run --machine --start-paused -d chrome --devtools-server-address http://127.0.0.1:9100/ --target <redacted>/<redacted>.dart`

Stack trace:
```
UnsupportedError: Unsupported operation: Illegal character in path: http:
at _Uri._checkWindowsPathReservedCharacters(uri.dart:1840)
at _Uri._toWindowsFilePath(uri.dart:2893)
at _Uri.toFilePath(uri.dart:2865)
at ManifestAssetBundle._parseAssetsFromFolder(asset.dart:895)
at ManifestAssetBundle._parseAssets(asset.dart:810)
at ManifestAssetBundle.build(asset.dart:265)
at <asynchronous gap>(async)
at ResidentWebRunner._updateDevFS(resident_web_runner.dart:527)
at <asynchronous gap>(async)
at ResidentWebRunner.run.<anonymous closure>(resident_web_runner.dart:320)
at <asynchronous gap>(async)
at asyncGuard.<anonymous closure>(async_guard.dart:111)
at <asynchronous gap>(async)
```

Up to date stack traces as of 29db9dbd44b1dbf781a84318811db676aadd6fd7 (using `flutter run -d chrome`):

<details>
<summary> "http://whitehouse.gov/foo.png" </summary>

```

FileSystemException: FileSystemException: Exists failed, path = 'C:\Users\Jozz\Desktop\deleteme\assets_test\http:\\whitehouse.gov' (OS Error: The filename, directory name, or volume label syntax is incorrect.
, errno = 123)

#0      _Directory.existsSync (dart:io/directory_impl.dart:87:7)
#1      ForwardingFileSystemEntity.existsSync (package:file/src/forwarding/forwarding_file_system_entity.dart:43:33)
#2      ErrorHandlingDirectory.existsSync.<anonymous closure> (package:flutter_tools/src/base/error_handling_io.dart:490:22)
#3      _runSync (package:flutter_tools/src/base/error_handling_io.dart:587:14)
#4      ErrorHandlingDirectory.existsSync (package:flutter_tools/src/base/error_handling_io.dart:489:12)
#5      _AssetDirectoryCache.variantsFor (package:flutter_tools/src/asset.dart:1254:43)
#6      ManifestAssetBundle._parseAssetFromFile (package:flutter_tools/src/asset.dart:974:37)
#7      ManifestAssetBundle._parseAssets (package:flutter_tools/src/asset.dart:829:9)
#8      ManifestAssetBundle.build (package:flutter_tools/src/asset.dart:267:54)
<asynchronous suspension>
#9      ResidentWebRunner._updateDevFS (package:flutter_tools/src/isolated/resident_web_runner.dart:531:26)
<asynchronous suspension>
#10     ResidentWebRunner.run.<anonymous closure> (package:flutter_tools/src/isolated/resident_web_runner.dart:319:41)
<asynchronous suspension>
#11     asyncGuard.<anonymous closure> (package:flutter_tools/src/base/async_guard.dart:111:24)
<asynchronous suspension>
```

</details>


<details>
<summary> "http://whitehouse.gov/pictures/" </summary>

```

UnsupportedError: Unsupported operation: Illegal character in path: http:

#0      _Uri._checkWindowsPathReservedCharacters (dart:core/uri.dart:1840:11)
#1      _Uri._toWindowsFilePath (dart:core/uri.dart:2893:7)
#2      _Uri.toFilePath (dart:core/uri.dart:2865:38)
#3      ManifestAssetBundle._parseAssetsFromFolder (package:flutter_tools/src/asset.dart:916:29)
#4      ManifestAssetBundle._parseAssets (package:flutter_tools/src/asset.dart:817:9)
#5      ManifestAssetBundle.build (package:flutter_tools/src/asset.dart:267:54)
<asynchronous suspension>
#6      ResidentWebRunner._updateDevFS (package:flutter_tools/src/isolated/resident_web_runner.dart:531:26)
<asynchronous suspension>
#7      ResidentWebRunner.run.<anonymous closure> (package:flutter_tools/src/isolated/resident_web_runner.dart:319:41)
<asynchronous suspension>
#8      asyncGuard.<anonymous closure> (package:flutter_tools/src/base/async_guard.dart:111:24)
<asynchronous suspension>
```
</details>

Another thing to note is that if I instead use `flutter run -d windows`, I don't get a crash. Instead, I get an error from VS that's even less useful than a crash:

```
Launching lib\main.dart on Windows in debug mode...
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(249,5): error MSB8066: Custom build for 'C:\Users\Jozz\Desktop\deleteme\assets_test\build\windows\x64\CMakeFiles\2103228256e52c9a1428711c685a3aa3\flutter_windows.dll.rule;C:\Users\Jozz\Desktop\deleteme\assets_test\build\windows\x64\CMakeFiles\b847e1c2ab5b4b4fa16f57a4fdf59571\flutter_assemble.rule' exited with code 1. [C:\Users\Jozz\Desktop\deleteme\assets_test\build\windows\x64\flutter\flutter_assemble.vcxproj]
Building Windows application...                                     5.9s
Error: Build process failed.
```

--- a/packages/flutter_tools/lib/src/asset.dart
+++ b/packages/flutter_tools/lib/src/asset.dart
@@ -918,7 +918,7 @@ class ManifestAssetBundle implements AssetBundle {
           .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
       } on UnsupportedError catch (e) {
         throwToolExit(
-          'Unable to search for asset files in directory path "${Uri.decodeFull(assetUri.path)}". '
+          'Unable to search for asset files in directory path "${assetUri.path}". '
           'Please ensure that this is valid URI that points to a directory '
           'that is available on the local file system.\nError details:\n$e');
       }

--- a/packages/flutter_tools/lib/src/asset.dart
+++ b/packages/flutter_tools/lib/src/asset.dart
@@ -912,8 +912,17 @@ class ManifestAssetBundle implements AssetBundle {
     Package? attributedPackage,
     List<String>? flavors,
   }) {
-    final String directoryPath = _fileSystem.path.join(
-        assetBase, assetUri.toFilePath(windows: _platform.isWindows));
+    final String directoryPath = (() {
+      try {
+        return _fileSystem.path
+          .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
+      } on UnsupportedError catch (e) {
+        throwToolExit(
+          'Unable to search for asset files in directory path "${Uri.decodeFull(assetUri.path)}". '
+          'Please ensure that this is valid URI that points to a directory '
+          'that is available on the local file system.\nError details:\n$e');
+      }
+    })();
 
     if (!_fileSystem.directory(directoryPath).existsSync()) {
       _logger.printError('Error: unable to find directory entry in pubspec.yaml: $directoryPath');
@@ -1251,7 +1260,7 @@ class _AssetDirectoryCache {
   List<String> variantsFor(String assetPath) {
     final String directory = _fileSystem.path.dirname(assetPath);
 
-    if (!_fileSystem.directory(directory).existsSync()) {
+    if (!_fileSystem.directory(directory).existsSync()) { // HERE
       return const <String>[];
     }
 

--- a/packages/flutter_tools/lib/src/asset.dart
+++ b/packages/flutter_tools/lib/src/asset.dart
@@ -912,17 +912,16 @@ class ManifestAssetBundle implements AssetBundle {
     Package? attributedPackage,
     List<String>? flavors,
   }) {
-    final String directoryPath = (() {
-      try {
-        return _fileSystem.path
-          .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
-      } on UnsupportedError catch (e) {
-        throwToolExit(
-          'Unable to search for asset files in directory path "${assetUri.path}". '
-          'Please ensure that this is valid URI that points to a directory '
-          'that is available on the local file system.\nError details:\n$e');
-      }
-    })();
+    final String directoryPath;
+    try {
+      directoryPath = _fileSystem.path
+        .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
+    } on UnsupportedError catch (e) {
+      throwToolExit(
+        'Unable to search for asset files in directory path "${assetUri.path}". '
+        'Please ensure that this is valid URI that points to a directory '
+        'that is available on the local file system.\nError details:\n$e');
+    }
 
     if (!_fileSystem.directory(directoryPath).existsSync()) {
       _logger.printError('Error: unable to find directory entry in pubspec.yaml: $directoryPath');

--- a/packages/flutter_tools/lib/src/asset.dart
+++ b/packages/flutter_tools/lib/src/asset.dart
@@ -1260,7 +1260,7 @@ class _AssetDirectoryCache {
   List<String> variantsFor(String assetPath) {
     final String directory = _fileSystem.path.dirname(assetPath);
 
-    if (!_fileSystem.directory(directory).existsSync()) { // HERE
+    if (!_fileSystem.directory(directory).existsSync()) {
       return const <String>[];
     }
 

--- a/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
+++ b/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
@@ -171,14 +171,9 @@ flutter:
         'Unsupported operation: Illegal character in path: https:',
       ));
     }, overrides: <Type, Generator>{
-      FileSystem: () {
-        final FileSystem result = MemoryFileSystem(
-          style: FileSystemStyle.windows,
-        );
-        result.currentDirectory = result.systemTempDirectory.createTempSync('flutter_asset_bundle_test.');
-        return result;
-      },
+      FileSystem: () => testFileSystem,
       ProcessManager: () => FakeProcessManager.any(),
+      Platform: () => FakePlatform(operatingSystem: 'windows'),
     });
 
     testUsingContext('handle removal of wildcard directories', () async {

--- a/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
+++ b/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
@@ -153,8 +153,8 @@ flutter:
     });
 
     testUsingContext('throws ToolExit when directory entry contains invalid characters', () async {
-      globals.fs.file('.packages').createSync();
-      globals.fs.file('pubspec.yaml')
+      testFileSystem.file('.packages').createSync();
+      testFileSystem.file('pubspec.yaml')
         ..createSync()
         ..writeAsStringSync(r'''
 name: example

--- a/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
+++ b/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
@@ -163,7 +163,7 @@ flutter:
     - https://mywebsite.com/images/
 ''');
       final AssetBundle bundle = AssetBundleFactory.instance.createBundle();
-      expect (() => bundle.build(packagesPath: '.packages'), throwsToolExit(
+      expect(() => bundle.build(packagesPath: '.packages'), throwsToolExit(
         message: 'Unable to search for asset files in directory path "https://mywebsite.com/images/". '
         'Please ensure that this is valid URI that points to a directory that is '
         'available on the local file system.\n'
@@ -171,7 +171,13 @@ flutter:
         'Unsupported operation: Illegal character in path: https:',
       ));
     }, overrides: <Type, Generator>{
-      FileSystem: () => testFileSystem,
+      FileSystem: () {
+        final FileSystem result = MemoryFileSystem(
+          style: FileSystemStyle.windows,
+        );
+        result.currentDirectory = result.systemTempDirectory.createTempSync('flutter_asset_bundle_test.');
+        return result;
+      },
       ProcessManager: () => FakeProcessManager.any(),
     });
 

--- a/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
+++ b/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
@@ -152,6 +152,29 @@ flutter:
       ProcessManager: () => FakeProcessManager.any(),
     });
 
+    testUsingContext('throws ToolExit when directory entry contains invalid characters', () async {
+      globals.fs.file('.packages').createSync();
+      globals.fs.file('pubspec.yaml')
+        ..createSync()
+        ..writeAsStringSync(r'''
+name: example
+flutter:
+  assets:
+    - https://mywebsite.com/images/
+''');
+      final AssetBundle bundle = AssetBundleFactory.instance.createBundle();
+      expect (() => bundle.build(packagesPath: '.packages'), throwsToolExit(
+        message: 'Unable to search for asset files in directory path "https://mywebsite.com/images/". '
+        'Please ensure that this is valid URI that points to a directory that is '
+        'available on the local file system.\n'
+        'Error details:\n'
+        'Unsupported operation: Illegal character in path: https:',
+      ));
+    }, overrides: <Type, Generator>{
+      FileSystem: () => testFileSystem,
+      ProcessManager: () => FakeProcessManager.any(),
+    });
+
     testUsingContext('handle removal of wildcard directories', () async {
       globals.fs.file(globals.fs.path.join('assets', 'foo', 'bar.txt')).createSync(recursive: true);
       final File pubspec = globals.fs.file('pubspec.yaml')

--- a/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
+++ b/packages/flutter_tools/test/general.shard/asset_bundle_test.dart
@@ -164,7 +164,7 @@ flutter:
 ''');
       final AssetBundle bundle = AssetBundleFactory.instance.createBundle();
       expect(() => bundle.build(packagesPath: '.packages'), throwsToolExit(
-        message: 'Unable to search for asset files in directory path "https://mywebsite.com/images/". '
+        message: 'Unable to search for asset files in directory path "https%3A//mywebsite.com/images/". '
         'Please ensure that this is valid URI that points to a directory that is '
         'available on the local file system.\n'
         'Error details:\n'

